backend/cmm_helpers.ml

```diff
@@ -2449,24 +2449,34 @@ let has_local_allocs e =
     | Cop (Calloc Alloc_local, _, _) | Cop ((Cextcall _ | Capply _), _, _) ->
       raise Exit
     | e -> iter_shallow loop e
-  and loop_until_tail = function
-    | Ctail e -> loop e
-    | Cregion _ -> ()
-    | e -> ignore (iter_shallow_tail loop_until_tail e)
+  and loop_until_tail e = loop_until_tail0 e ~depth:0
+  and loop_until_tail0 ~depth = function
+    | Ctail e ->
+      if depth = 0 then loop e else loop_until_tail0 ~depth:(depth - 1) e
+    | Cregion e -> loop_until_tail0 ~depth:(depth + 1) e
+    | e -> ignore (iter_shallow_tail (loop_until_tail0 ~depth) e)
   in
   match loop e with () -> false | exception Exit -> true
```

Comment:
I think `has_local_allocs (Ctail (Cop (Calloc Alloc_local, _, _)))` should return `false`: the check is about whether it has local allocations in the *current* region. (In particular, `cregion` applied to the above should delete the `Ctail`, not add another `Cregion`).

So, how about the following implementation instead?
```ocaml
let rec loop ~depth = function
  | Cregion e -> loop ~depth:(depth+1) e
  | Ctail e -> if depth = 0 then () else loop ~depth:(depth - 1)
  | Cop (Calloc Alloc_local, _, _) | Cop ((Cextcall _ | Capply _), _, _) 
     when depth = 0 ->
    raise Exit
  | e -> iter_shallow loop e
```
```

Reply:
Yeah, works for me.